quinn-proto/src/connection/state.rs

```diff
@@ -0,0 +1,259 @@
+use bytes::Bytes;
+
+use crate::frame::Close;
+use crate::{ConnectionError, TransportErrorCode};
+
+#[allow(unreachable_pub)] // fuzzing only
+#[derive(Debug, Clone)]
+pub struct State {
+    inner: InnerState,
+}
+
+impl State {
+    pub(super) fn as_handshake_mut(&mut self) -> Option<&mut Handshake> {
+        if let InnerState::Handshake(ref mut hs) = self.inner {
+            Some(hs)
+        } else {
+            None
+        }
+    }
+
+    pub(super) fn as_handshake(&self) -> Option<&Handshake> {
+        if let InnerState::Handshake(ref hs) = self.inner {
+            Some(hs)
+        } else {
+            None
+        }
+    }
+
+    pub(super) fn as_closed(&self) -> Option<&Closed> {
+        if let InnerState::Closed {
+            ref remote_reason, ..
+        } = self.inner
+        {
+            Some(remote_reason)
+        } else {
+            None
+        }
+    }
+
+    #[allow(unreachable_pub)] // fuzzing only
+    #[cfg(test)]
+    pub fn established() -> Self {
+        Self {
+            inner: InnerState::Established,
+        }
+    }
+
+    pub(super) fn handshake(handshake: Handshake) -> Self {
+        Self {
+            inner: InnerState::Handshake(handshake),
+        }
+    }
+
+    pub(super) fn to_handshake(&mut self, hs: Handshake) {
```

Comment:
This kind of clashes with the rust conventions. I'd suggest `transition_to_handshake()` maybe?

Reply:
done